File: tang/src/main/java/com/microsoft/tang/examples/timer/Timer.java
package com.microsoft.tang.examples.timer;

import com.microsoft.tang.annotations.DefaultImplementation;
import com.microsoft.tang.annotations.Name;
import com.microsoft.tang.annotations.NamedParameter;

@DefaultImplementation(TimerImpl.class)
public interface Timer {
  @NamedParameter(default_value="10",
      doc="Number of seconds to sleep", short_name="sec")
  public static class Seconds implements Name<Integer> { }
  public void sleep() throws Exception;
}

File: tang/src/main/cs/Examples/Timer.cs
using System;
using System.Threading;
using Com.Microsoft.Tang.Annotations;

namespace Com.Microsoft.Tang.Examples
{
    public class Timer
    {
        [NamedParameter("Number of seconds to sleep", "sec", "10")]
        class Seconds : Name<Int32> { }
        private readonly int seconds;

        [Inject]
        public Timer([Parameter(Value = typeof(Seconds))] int seconds)
        {
            if (seconds < 0)
            {
                throw new ArgumentException("Cannot sleep for negative time!");
            }
            this.seconds = seconds;
        }

        public void sleep()  
        {
            Thread.Sleep(seconds * 1000);
        }
    }
}
